### DISCUSSION TITLE
[Q] How to build a tree_sitter::Tree from a sexp representation?

### DISCUSSION THREAD

**Discussion by gagbo** [tree-sitter/tree-sitter#2293] [Q&A] [answered]
Hello,

I'd like to be able to somehow save and persist old parsed trees to reuse them in `Parser::parse` across my application restarts.

Since the `Tree` structure is a pointer to an FFI struct, I think it's impossible to use a serializer to save the binary data representing the tree to disk in order to retrieve it later.

My next idea is then to save the `tree.root_node().to_sexp()` (sorry I don't remember the API well) to disk, and then later read this to rebuild a tree directly out of it when needed. Is there a function that would allow doing exactly that? I was imagining something like `fn try_from_sexp<'source>(sexp: &str, source: &'source str) -> Result<Tree, _>` signature, that wouldn't reparse everything but just "restore" a parsed tree. Might even be an unsafe function, with the safety invariant being that the user of the function should manually ensure that the `source` still matches the `sexp`

Thanks for the lib!

**Comment by ahlinc** [comment, marked as answer]
Tree-sitter doesn't support such case.

1. Tree-sitter's trees are much complex internally and the S-Expression representation shows just a part.
2. If it would be possible to serialize the tree it would take much more space on the disk that original text.

So for the first parse it's better to store an original text and may be in a compressed form.

**Reply by gagbo** [reply]
I see, that makes sense I guess.

FTR I was thinking of serializing a struct like the data showed by the [TreeSitter playground from neovim](https://user-images.githubusercontent.com/2361214/202389106-244ac890-9442-4759-9b2c-4fe3c247dfbc.gif) (so node type and the byte ranges basically). And even if it ends up pretty big, I thought that reading and restoring trees from this information would be better than re-parsing a big `.org` file.

Thanks for the answer :)
